File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0

def dfs(index, total, numbers, target):
    global answer 

    if index == len(numbers) and total == target:        
        answer += 1
        return
    if index == len(numbers):
        return
    
    dfs(index+1, total + numbers[index], numbers, target)
    dfs(index+1, total - numbers[index], numbers, target)

def solution(numbers, target):    
    dfs(0, 0, numbers, target)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;

    public int solution(int[] numbers, int target) {
        dfs(0, 0, numbers, target);
        return answer;
    }

    void dfs(int index, int total, int[] numbers, int target) {
        if (index == numbers.length && total == target) {
            answer++;
            return;
        }
        if (index == numbers.length) {
            return;
        }

        dfs(index + 1, total + numbers[index], numbers, target);
        dfs(index + 1, total - numbers[index], numbers, target);
    }
}
